### PR TITLE
imx6ull: nor-driver qspi

### PR DESCRIPTION
**Pull request by badochov**
JIRA: RTOS-319

## Description
imx6ull-flashnor qspi driver
This driver is almost copy paste of flash-flexspi with fspi module changed to qspi.

## Motivation and Context
imx6ull lacks proper plo, flash driver is a needed milestone to implement it.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: imx6ull.

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Comment by agkaminski** [comment]
@badochov yes, if the code is reasonably lightweight, it can even be included in the standard command set (for appropriate targets)

File: devices/flash-imx6ull/qspi/qspi.c [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+
+	/* Lock LUT */
+	QSPI_BASE[QSPI_LUTKEY] = QSPI_LUTKEY_KEY;
+	QSPI_BASE[QSPI_LCKCR] = (QSPI_BASE[QSPI_LCKCR] & ~(QSPI_LCKCR_UNLOCK | QSPI_LCKCR_LOCK)) | QSPI_LCKCR_LOCK;
+}
+
+
+__attribute__((section(".noxip"))) static void fspi_setMux(int dev_no)
+{
+	unsigned int i;
```
Comment (agkaminski) [inline devices/flash-imx6ull/qspi/qspi.c:108]:
```suggestion
	size_t i;
```

File: devices/flash-imx6ull/qspi/qspi.c [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+		case 3:
+			reg = QSPI_BASE[QSPI_SFB1AD];
+			break;
+		default:
+			reg = 0;
+			break;
+	}
+	reg = (reg >> 10) << 10;
+
+	return addr + reg;
```
Comment (agkaminski) [inline devices/flash-imx6ull/qspi/qspi.c:254]:
```suggestion
	return addr + (reg & (~0x3ffu));
```

File: devices/flash-imx6ull/qspi/qspi.c [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+	switch (xfer->op) {
+		case xfer_opWrite: /* IP write is limited to IPDATSZ mask */
+			if (xfer->data.write.sz > 0xffffu) {
+				return -EPERM;
+			}
+			res = qspi_opWrite(qspi, start, xfer);
+			break;
+		case xfer_opRead:
+			/* For >64k read out the data directly from the AHB buffer (data may be cached) */
+			/* TODO Test */
```
Comment (agkaminski) [inline devices/flash-imx6ull/qspi/qspi.c:383]: Has it been tested after all?

**Review by agkaminski — COMMENTED** [review]
Nitpicks only

**Comment by nalajcie** [comment]
Shouldn't we name in `imx6ull-flashnor` to be consistent with `phoenix-rtos-devices`? (And I would like to avoid `imx6ull-flashnand` in the future :))

**Comment by badochov** [comment]
> Shouldn't we name in `imx6ull-flashnor` to be consistent with `phoenix-rtos-devices`? (And I would like to avoid `imx6ull-flashnand` in the future :))

If I understand PLO driver structure correctly there has to be a common driver for both nor and nand hence the name

**Comment by gerard5** [comment]
I don't understand why duplicate the`nor.c` implementation when the only thing that has changed there is `flexspi_*` ->`qspi_*`. Originally, the idea was to make the `nor` memory driver common to the whole `i.MX` family (`flashdrv.c`). Both `flexspi` and `qspi` drivers are family (90% compatibile). Having one common implementation of `nor.c` all targets in the `i.MX` family benefit and the development is simpler (`DRY` principle). Btw. the idea of `flexspi.c` was to easily add `dtb` config support, that's the purpose of setup structures in `*.h` files and `nor_probe` function of `flashdrv.c`, the underlying layer `flexspi` or `qspi` should be interchangeable.

File: devices/flash-imx6ull/qspi/qspi.c [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+
+__attribute__((section(".noxip"))) int qspi_init(qspi_t *qspi, u8 slPortMask)
+{
+	u32 lut[4];
+	unsigned i;
+	const size_t flashszs[4] = { 4 * 1024 * 1024, 4 * 1024 * 1024, 4 * 1024 * 1024, 4 * 1024 * 1024 };
+
+	if (((slPortMask & ~0xf) != 0) || ((slPortMask & 0xf) == 0)) {
+		return -EINVAL;
+	}
```
Comment (kemonats) [inline devices/flash-imx6ull/qspi/qspi.c:163]: Probably simpler
```suggestion
	if (((slPortMask > 15) || ((slPortMask == 0)) {
		return -EINVAL;
	}
```

File: devices/flash-imx6ull/qspi/qspi.c (hunk 3 above)
Comment (badochov) [reply devices/flash-imx6ull/qspi/qspi.c]: In current commit

**Review by agkaminski — APPROVED** [review]
LGTM

**Review by agkaminski — APPROVED** [review]

**Pull request merged**